src/lexer.rs

```diff
@@ -0,0 +1,432 @@
+use crate::{
+    asciifile::{Position, PositionedChar, PositionedChars},
+    strtab::*,
+};
+
+use std::{convert::TryFrom, result::Result};
+
+macro_rules! match_op {
+    ($input: expr, $len: expr, $right: expr) => {{
+        // Unwraps are safe, because this is only called after token of length $len,
+        // is already matched and thus contained in $input
+        let PositionedChar(begin, _) = $input.next().unwrap();
+        let mut end = begin;
+        #[allow(clippy::reverse_range_loop)] // Macro might be called with with $len=1
+        for _ in 1..$len {
+            let PositionedChar(pos, _) = $input.next().unwrap();
+            end = pos;
+        }
+
+        Some(Token::new(begin, end, TokenData::Operator($right)))
+    }};
+}
+
+#[derive(Debug)]
+pub enum TokenData<'t> {
+    Keyword(Keyword),
+    Operator(Operator),
+    Identifier(Symbol<'t>),
+    IntegerLiteral(Symbol<'t>),
+    Comment(String),
+    UnclosedComment(String),
+    UnexpectedCharacter(char),
```

Comment:
This was easier in the beginning, which is why I went for it. It would probably be nicer to make the lexer an iterator over `Result<Token, LexicalError>`, where `LexicalError` contains variants for unclosed comments & unexpected characters.